bentoml/entrypoint/__init__.py

```diff
@@ -0,0 +1,133 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import multiprocessing
+from typing import Optional
+
+from bentoml.configuration.containers import BentoMLConfiguration, BentoMLContainer
+from bentoml.utils import reserve_free_port
+
+logger = logging.getLogger(__name__)
+
+
+def start_prod_server(
+    saved_bundle_path: str,
+    port: Optional[int] = None,
+    workers: Optional[int] = None,
+    timeout: Optional[int] = None,
+    enable_microbatch: Optional[bool] = None,
+    enable_swagger: Optional[bool] = None,
+    mb_max_batch_size: Optional[int] = None,
+    mb_max_latency: Optional[int] = None,
+    microbatch_workers: Optional[int] = None,
+    config_file: Optional[str] = None,
+):
+    import psutil
+
+    assert (
+        psutil.POSIX
+    ), "BentoML API Server production mode only supports POSIX platforms"
+
+    config = BentoMLConfiguration(override_config_file=config_file)
+    config.override(["api_server", "port"], port)
+    config.override(["api_server", "workers"], workers)
+    config.override(["api_server", "timeout"], timeout)
+    config.override(["api_server", "enable_microbatch"], enable_microbatch)
+    config.override(["api_server", "enable_swagger"], enable_swagger)
+    config.override(["marshal_server", "max_batch_size"], mb_max_batch_size)
+    config.override(["marshal_server", "max_latency"], mb_max_latency)
+    config.override(["marshal_server", "workers"], microbatch_workers)
+
+    if config.config['api_server'].get('enable_microbatch'):
+        prometheus_lock = multiprocessing.Lock()
+        with reserve_free_port() as api_server_port:
+            pass
+
+        model_server_job = multiprocessing.Process(
+            target=_start_prod_server,
+            kwargs=dict(
+                saved_bundle_path=saved_bundle_path,
+                port=api_server_port,
+                config=config,
+                prometheus_lock=prometheus_lock,
+            ),
+            daemon=True,
+        )
+        model_server_job.start()
+
+        try:
+            _start_prod_batching_server(
+                saved_bundle_path=saved_bundle_path,
+                config=config,
+                api_server_port=api_server_port,
+                prometheus_lock=prometheus_lock,
+            )
+        finally:
+            model_server_job.terminate()
+    else:
+        _start_prod_server(saved_bundle_path=saved_bundle_path, config=config)
+
+
+def _start_prod_server(
+    saved_bundle_path: str,
+    config: BentoMLConfiguration,
+    port: Optional[int] = None,
```

Comment:
Seeing the batching app and the model app as the whole API server, it makes sense to keep `config.api_server` rather than override it to the randomly picked port.

Reply:
I understand the concern. Here is why we should restructure the config keys and terminology. Before we do that, I think there is an opportunity here to make the intermediate port injectable as well. First we can introduce an intermediate port (for lack of a better name).

```
api_server:
    marshal:
        intermediate_port: Null # or an actual port e.g. 6000
```

In the container, we can introduce a provider that first checks if there is an intermediate port defined, if not randomly reserve a port.

```
intermediate_port = providers.Callable(
    lambda port: port if port else reserve_free_port(),
    config.marshal.intermediate_port,
)
```

Reply:
It's great to have `reserve_free_port ` as a provider of the `intermediate_port`. But we have to wire after creating the new process. Then the `reserve_free_port` would be called twice and gave `intermediate_port` different values in each process.

Reply:
Good call that `reserve_free_port` might be called twice. But it should be a solvable problem. We can use a singleton provider for creating the intermediate port. All users of the provider will get the same port. We will need to move the container creation out, however. Happy to discuss with you over a Zoom call.

Reply:
I suggest using UDS in the future instead of introducing a new config that users don't need to know.

Reply:
UDS is superior but we might have to leave the TCP option open to make remote marshaling a possibility. We can structure the config meaningfully to reflect this.

Reply:
> to leave the TCP option open to make remote marshaling a possibility. We can structure the config meaningfully to reflect this.
```
api_server:
  port: 5000
  enable_microbatch: False
  run_with_ngrok: False
  enable_swagger: True
  enable_metrics: True
  enable_feedback: True
  max_request_size: 20971520
  workers: 1
  timeout: 60

model_server:
  port: Null  # default: api_server.port when enable_marshal=False

marshal_server:
  port: Null  # default: api_server.port when enable_marshal=True


```
like this?

Reply:
I'm thinking something like the following where the using can choose the connector type of provide related configs. Basically the schema for connector is a union of UDS and TCP connector schemas.

If UDS is chosen,
```
api_server:
  marshal:
    connector:
      type: UDS
      uds_related_key1: value1
      uds_related_key2: value2
```
Or, if TCP is chosen,
```
api_server:
  marshal:
    connector:
      type: TCP
      port: None # or some configured port
      tcp_related_key1: value1
```

Reply:
For UDS support, I'd like to have a simple `host` field in URI format, just like gunicorn and Nginx does.
- `127.0.0.1:5000`
- `unix:/tmp/gunicorn.sock`

Your schema is fancy but looks too powerful for me.
I think you can draft a new PR to demonstrate your idea.

Reply:
After all, the config structure reflects our system design. We can continue the discussion in the new PR/issue.